### DISCUSSION TITLE
'standardize' menus

### DISCUSSION THREAD

**Discussion by jp8** [jamulussoftware/jamulus#1915] [UI Related]
A recurring theme in the github discussions is that users do not easily find the functionality in the menus.

I have put the word 'standardize' in quotes in the topic, because it is not possible for everyone to agree on what is standard.  However the screenshots I am attaching are at least closer to what I personally perceive as common usage.

![Screenshot_2021-07-14_19-05-05](https://user-images.githubusercontent.com/6978815/125671776-07039d9a-1f03-49fa-899a-8088bd23737d.png)
![Screenshot_2021-07-14_19-05-17](https://user-images.githubusercontent.com/6978815/125671786-12532f3d-13cd-4d7c-8cdd-12fd38c347aa.png)
![Screenshot_2021-07-14_19-05-20](https://user-images.githubusercontent.com/6978815/125671789-3a93c841-5f22-42d3-8f18-25e383d9078a.png)
![Screenshot_2021-07-14_19-05-26](https://user-images.githubusercontent.com/6978815/125671798-46fccfbc-bf1d-49fa-80db-f7c3ee70c06c.png)
![Screenshot_2021-07-14_19-05-29](https://user-images.githubusercontent.com/6978815/125671802-7b707eef-6e4d-403e-ae37-7a643b8ecf26.png)
![Screenshot_2021-07-14_19-05-35](https://user-images.githubusercontent.com/6978815/125671811-bb97a5c0-3e93-4a88-bca6-74c4a1ae640b.png)

The advantage is:
- users will more easily find what they are looking for (at least, that's the goal)

The obvious disadvantages are:
- existing guides and documentation will need updating
- might confuse or surprise existing users

My view is that even existing users would prefer the new layout.

**Comment by dcorson-ticino-com** [comment]
I agree with you that that would be better, especially for new users.
I wonder 'though if it would not be better to name "View" to "Sort" as all the entries have to do with sorting and nothing else.
Of course it is not a standard menu name, but it would at least be explicit.

**Reply by jp8** [reply]
The reason I used 'view' is:
- it seemed like the standard, expected word
- i would at some point like to put the  compact | fancy | normal  choice there, instead of its current location (my profile)

![Screenshot_2021-07-15_12-23-49](https://user-images.githubusercontent.com/6978815/125780406-98c4471b-4994-4363-9472-3c8dbedee635.png)

**Comment by mulyaj** [comment]
I've actually been running a card sort on this topic for around a couple months now, but it still has low responses. It's the [one on the announcements](https://github.com/jamulussoftware/jamulus/discussions/1858). I just need around 10-20 more responses for some usable data on how the menu should be sorted.

**Reply by jp8** [reply]
OK I have responded to your survey.

**Comment by henkdegroot** [comment]
I like the idea of this change, though some menu items do not really make sense even with the current menu it is not making sense. I.e. in a menu named "View" you would not expect it will provide access to different dialogs related to (mainly) settings. And "Edit" is not really allowing you to edit. So definitely improvement can be done.

In the proposal, there still is the "Edit" menu which does not "edit". I believe these items fall into the same category as "Settings".

Not sure how self-explanatory the "Window" menu is. If I would need to access the "Connection Setup" it would not be the first menu I would check. It is the only dialog that cannot be directly accessed from the main GUI when you are in an active session (have already connected). Perhaps that is something which should be improved?

**Reply by jp8** [reply]
Here are some of the principles I tried to use when categorizing:

About the EDIT menu:

If you think of the Jamulus UI as a mixing console, then editing is the action of moving faders, muting and soloing channels.  That is why I put all these on the Edit menu.  If there was a copy and paste in Jamulus, it would, I suppose, be to copy one fader setting and apply it to another fader or group of faders.  Undo and Redo, if they existed, would belong here, and would undo any fader changes.

This is functionality that woud be used while jamming.

About the SETTINGS menu:

This is for the user's own personal settings.  Their input gain, their audio device buffer configuration, their name and instrument.  Everything on here would normally be configured before the user starts jamming.

**Reply by henkdegroot** [reply]
I don't want to be a pain, but if you need to explain or way think of....it would deveat the intended purpose of making things more obvious to a first time/less experienced user.

If all items in the edit menu are only related to the mixerboard, it could be call mixer options?
Eventhougjh that might be more restricing in the future.

**Reply by jp8** [reply]
Not at all.

You're right there is an argument for putting those functions on an 'application specific' menu, and naming it something like 'mixer options'.

But I think there's a stronger argument for the Edit menu being the correct place.

Apple has done a reasonable job of documenting the design principles for application menus, and their conventions are typically followed outside Apple as well:

https://developer.apple.com/design/human-interface-guidelines/macos/menus/menu-bar-menus/

Note the conventions on Windows and Unix differ slightly from Apple's.  An Apple user would be surprised by the Settings menu, and would instead expect a single 'Preferences' item under the 'Jamulus' menu.  But Windows and Unix would not even have a 'Jamulus' menu, so it has to go somewhere else.

**Reply by hoffie** [reply]
> I don't want to be a pain, but if you need to explain or way think of....it would deveat the intended purpose of making things more obvious to a first time/less experienced user.

One might argue that such an explanation helps keeping this (long-term and sometimes heated) discussion focused and ensures that future iterations don't inadvertently revert an intended change here? :)

**Comment by hoffie** [comment]
I really like the proposal. If feels way more intuitive than the status quo. We should, however, try to cross-check this proposal with the data from @mulyaj's survey (which I've just filled as well).

**Comment by ghost** [comment]
Everything can fit under the following categories:
File Settings Preferences Help

In UNIX almost everything is a File including files, audio and internet connections.
Settings are settings (possibly placed in a file).
Preferences are personal choices (possibly placed in a file).
Help contains information (read from a file).

**Comment by gilgongo** [comment]
> In UNIX almost everything is a File including files, audio and internet connections.

We do need to remember that less than 5% of Jamulus users are on *nix though :-)

**Reply by jp8** [reply]
No matter what platform, the point of the UI is to hide that level of detail from the user.

**Reply by ghost** [reply]
> No matter what platform, the point of the UI is to hide that level of detail from the user.

... without adding abstract clutter from an arcane paradigm (eg. GIMP)
KISS

**Reply by jp8** [reply]
Precisely.

**Comment by jp8** [comment]
I'm interested to see the results of the survey but I'm not sure how long I'd like to wait, for 10-20 more submissions - especially considering the submissions won't include any justification for the design choices.

@mulyaj is there anything in the submissions so far that is significantly different from what we're discussing ?

**Reply by mulyaj** [reply]
Right now there are 13 submissions, I'd need about 20-30 to get a statistically significant structure. 

From the data now, the most obvious grouping is the Help section, which is pretty much the same as current Jamulus. The item grouping that has the most naming variance are Load Mixer, Save Mixer, Clear All Stored... and Set All Faders.... It looks like most people have really latched on to the mixer analogy. i.e Mixer, Jamming Controls, Mixer Settings, Mixer Management. In addition, the UI stuff like "Use Two Rows..." and "Sort" also bleeds into the "Mixer" grouping as well.

"File" and "Edit" also don't appear often either. File only has the grouping for "Load.. + Save.. Mixer Channels", with three submissions. 

Here's the dendrogram and a plot of the items against groups. If you'd like the raw data let me know. 
![image](https://user-images.githubusercontent.com/63881221/126347311-346fc366-fe27-4d35-bbc8-de35d4fb3bfb.png)

![image](https://user-images.githubusercontent.com/63881221/126347508-55598198-6a00-447a-ab7d-d666227c774b.png)

**Reply by jp8** [reply]
Ok so if I understand it all correctly, the main topics for discussion would be:

1 - naming of Edit menu  - should it be named something to do with Mixer ?
2 - separate menus for Settings and Window, or combine them.

To me, the 'standard' name Edit will be more familiar to users.

The responses indicate the people think Connection Setup and My Profile should be together on the Settings menu, but to me it doesn't make sense.  The Connect dialog is very different from settings.  Is the Chat menu item missing from the survey ?

IBM's CUA guidelines would place connection functionality on the File menu (possibly renaming it to the Connect menu) but in my view those guidelines are too outdated.  I think it is quite uncommon now to rename the File menu.

**Reply by mulyaj** [reply]
I think connection setup is just poorly named, which is why people place it in the Settings menu. From first glance, the action sounds like "settings for connection and other network- related features", but in reality, it's just "Connect to a Server".

**Reply by jp8** [reply]
Perhaps something like this:

File | Mix | View | Connect | Bookmarks | Settings | Help

With Chat as a 'checkbox' menu item on the View menu.  I added bookmarks just because I imagine it will be there at some point in the future.

And a test with a few users to see if Edit or Mix makes more sense to them.

**Reply by jp8** [reply]
Connection | Mix | View | Bookmarks | Settings | Help

File | Edit | View | Connection | Settings | Help

Are other possibilities.  I will try to mock up some screenshots.

**Reply by mulyaj** [reply]
Could also be Connect | Mixer w/ View Nested | Bookmarks | Settings | Help 

Using "Mixer" rather than "Mix", mix describes the total setup of the sound, while "mixer" describes the analogous object. 

The view within "Mixer" would contain the Mixer sorting options and "Use Two Rows".

**Reply by jp8** [reply]
I think the view menu is important enough to be at the top level.

**Reply by henkdegroot** [reply]
> "Use Two Rows".

This is no longer a menu option, as it changed into a setting which allows up to 8 rows.

**Reply by jp8** [reply]
![Screenshot_2021-07-20_21-38-06](https://user-images.githubusercontent.com/6978815/126392282-3ef5acd2-15e6-4a01-81ab-5cc9ca5ac237.png)

**Reply by henkdegroot** [reply]
Already mention in another comment, but repeat here....perhaps the GUI style options should be in a sub-menu as well? Just to make it more clear what these options are about.

**Comment by jp8** [comment]
![Screenshot_2021-07-20_19-17-00](https://user-images.githubusercontent.com/6978815/126375114-89bef8e4-2259-45a4-8455-d47d21f58af4.png)



![Screenshot_2021-07-20_18-57-24](https://user-images.githubusercontent.com/6978815/126372648-96a2a121-d8b5-4776-bf92-c59e92c66d8f.png)



![Screenshot_2021-07-20_18-51-00](https://user-images.githubusercontent.com/6978815/126371920-31be921e-768c-4261-ab8f-b75d8802687b.png)



![Screenshot_2021-07-20_19-21-28](https://user-images.githubusercontent.com/6978815/126375695-16173dfd-e8e1-4e2f-b8d7-78e2cd7009f1.png)

![Screenshot_2021-07-20_19-12-57](https://user-images.githubusercontent.com/6978815/126374522-83f384ec-86bb-4f08-933c-1fe225466363.png)

**Comment by jp8** [comment]
![Screenshot_2021-07-20_18-32-45](https://user-images.githubusercontent.com/6978815/126369640-cda3ff1f-a0ea-4c5c-a768-30c733909190.png)


I think I like 'File' better than 'Connection'

**Reply by henkdegroot** [reply]
I do not find the "file" menu a logical one for the connection. You normally use this when you need to switch servers while already connected to a directory server.

I would not add the "gui" modes into the menu either. Normal use would be a 1 off thing, not a constant change required  and makes it more challanging for more "gui" style changes to be added/used.

**Reply by jp8** [reply]
Connection is a more accurate description but File will be more familiar to users.

Have a look at the screenshot of my terminal window.  The File menu has nothing to do with files... it's just the de-facto standard name where you can open what you want to work on.

![Screenshot_2021-07-20_20-02-55](https://user-images.githubusercontent.com/6978815/126380934-89638847-6043-4def-80fd-c50322304a58.png)

It would be interesting to analyze some examples, if you know of any, of applications that have chosen a different label for the File menu.

**Reply by jp8** [reply]
Here's another example:

![Screenshot_2021-07-20_20-09-35](https://user-images.githubusercontent.com/6978815/126381809-d9fcc006-f781-4275-a20c-5417db17be1c.png)

The 'open workspace' item corresponds closely with what we do with Jamulus when we open a connection.

The 'open recent' item is exactly where we would put the same option on the Jamulus UI.

**Reply by jp8** [reply]
Apple's guideline, which is *much* more recent than the IBM one, says:   In a non-document-based app, the File menu can be renamed or eliminated.

I just wonder what the users would find the most intuitive.

**Reply by jp8** [reply]
> I would not add the "gui" modes into the menu either. Normal use would be a 1 off thing, not a constant change required

I always switch to compact when there are too many channels to fit my screen.

**Reply by henkdegroot** [reply]
> I always switch to compact when there are too many channels to fit my screen.

As a new user, would you understand by looking at the item in the "View" menu, that this option would change the GUI layout/size/style?
I have checked a few other programs and they seem to use a "sub-menu" for this. I.e. Skin in the menu and a sub-menu which contains the three options.

**Reply by henkdegroot** [reply]
> It would be interesting to analyze some examples, if you know of any, of applications that have chosen a different label for the File menu.

![vlc-media](https://user-images.githubusercontent.com/13550012/126438091-2fd231ce-5ec2-463b-a005-41ead3135faf.png)

**Reply by jp8** [reply]
> I have checked a few other programs and they seem to use a "sub-menu" for this. I.e. Skin in the menu and a sub-menu which contains the three options.

Yes it's a really good approach - if there were 4 skins I definitely would have done this.  But there is a recommendation in the guideline to avoid submenus wherever possible, so for just 3 skins, it seemed to me better to just put them directly on the View menu.

**Reply by henkdegroot** [reply]
> Yes it's a really good approach - if there were 4 skins

If this is the approach taken, then there are too many "sort" options in the menu and they should then be moved into a sub-menu.

Guess the "skin" options in the menu can be worded different to include the fact they are a "skin/style" option.

**Reply by jp8** [reply]
Turns out there is in fact one too many "sort" options :)

Yes I already added the word 'skin' to each of the skin options.

I am ok with a sub-menu or not for either of them.  The guideline says to avoid submenus because they are complex for some users, so there is a bit of a convention that 'advanced' parameters are more ok for submenus.  I'll make some more mock-ups soon so people can get an idea how it might look in both approaches.

**Reply by jp8** [reply]
![Screenshot_2021-07-21_20-19-25](https://user-images.githubusercontent.com/6978815/126547224-c3b20163-9177-444c-9646-bc978ed401fa.png)

**Comment by dcorson-ticino-com** [comment]
My impression is that all is getting much more complicated.
I see no need to have the all the directories in the file menu, what do they have to do with files?
Why should the GUI setup not be in Setup ?
Load mixer channels setup in edit?  why not in setup if it is a setup?

**Reply by jp8** [reply]
> My impression is that all is getting much more complicated.

Yes I added a bunch of stuff that is not in Jamulus but theoretically could be.  It helps me to think about how extensible and how 'correct' the design is.

> I see no need to have the all the directories in the file menu, what do they have to do with files?

Indeed, the discussion above is about whether to call it 'File' or 'Connection'.  Many many applications have things on the File menu that are not really about files - you can see in my screenshot above of my terminal window, literally *none* of the choices on the File menu is about files.

The screenshots are just to help think through the design.  Do you think it would be ok to have those options, if the menu was named 'Connection' ?  Or do you think it's wrong to offer the directories in the menus at all ?

> Why should the GUI setup not be in Setup ?

So far as I can tell, *all* the guidelines specify the View menu for that.

For example, from Apple:  "The View menu lets the user customize the appearance of an app’s windows."

> Load mixer channels setup in edit? why not in setup if it is a setup?

Indeed, why not in File, since we load them from a file ?

In those mock-ups, I put it on Edit for 2 reasons:

1. It changes the state of the mixer, which to me is exactly the Edit menu.
2. To see what people think of it.

**Reply by jp8** [reply]
![Screenshot_2021-07-20_20-45-01](https://user-images.githubusercontent.com/6978815/126386087-5c142658-a00f-4a1a-86a8-00e0e4e48258.png)

Another example of functionality that *theoretically* could exist.  If we offered users a list of recent server connections, I think users would expect to find it on the 'File' menu (the standards all call it the File menu even if it actually has a different name).

**Reply by jp8** [reply]
> Load mixer channels setup in edit? why not in setup if it is a setup?

I wonder if it really is a setup ?  What exactly is being loaded and saved here ?  Fader levels associated with client names ?

If so, I think it should be called Load Mix and Save Mix, it belongs on the File menu

**Reply by jp8** [reply]
![Screenshot_2021-07-21_20-24-26](https://user-images.githubusercontent.com/6978815/126547931-e2c0404b-f6f5-4204-8fd9-c098c69351da.png)

**Comment by jp8** [comment]
For those who prefer to see how just the existing Jamulus functionality would look under the re-arranged menus:
![Screenshot_2021-07-20_20-59-02](https://user-images.githubusercontent.com/6978815/126387762-bcc48f38-7de1-4f53-b514-5bff2046f726.png)
![Screenshot_2021-07-20_20-59-07](https://user-images.githubusercontent.com/6978815/126387774-b0aeee03-c7bc-42d5-a985-e7e08d2346ac.png)
![Screenshot_2021-07-20_20-59-10](https://user-images.githubusercontent.com/6978815/126387784-c5889adf-92ba-4411-bf3a-7d475ca94edb.png)
![Screenshot_2021-07-20_20-59-12](https://user-images.githubusercontent.com/6978815/126387789-b10edab2-f97a-44f4-8014-5fbd8215f7fe.png)


I'm hesitant to rename File to Connection because I have not seen many applications that rename the File menu.

I'm extremely hesitant to rename Edit to Mixer.

My gut feeling is when users see File, Edit, View, Settings, Help they will feel very very comfortable, as it mirrors closely all the other applications they have used.

**Reply by jp8** [reply]
I think I understand why the Edit menu should not be named Mixer.

Literally everything in Jamulus is about the mixer because Jamulus fundamentally is a mixer.  The client gives the user a UI to edit his or her own mix on the mixer.  The server is a digital mixer in the cloud.

It would be wrong to call one of the menus Mixer because in one way or another, *all* the menus are about the mixer.  These still seem ok to me:

File | Edit | View | Settings | Help
Connection | Edit | View | Settings | Help
Connection | Mix | View | Settings | Help

**Reply by gilgongo** [reply]
Another way of approaching this is not to have a File | Edit etc. menu at all, and just have a single branching menu containing all commands. That means the user doesn't need to think about which menu to open for any particular command from the start, and may afford more space to surface command names after that.

**Reply by jp8** [reply]
> Another way of approaching this is not to have a File | Edit etc. menu at all, and just have a single branching menu 

Yes Gnome UI guidelines use that approach, but so far at least, Microsoft and Apple did not follow.
(EDIT: I see the Chrome and Edge browsers use this approach.)
(EDIT EDIT:  Except on Mac, where it has a normal menubar, and the File menu has items that have nothing to do with files...)

https://developer.gnome.org/hig/stable/primary-menus.html.en

I think it's better to design for what is expected by the users on Windows and Mac.

**Reply by jp8** [reply]
Ah the KDE guidelines have a very nice description of this 

- A menu bar is mandatory for applications that have a very complex command structure , such as those used for content creation or editing, file manipulation, or other productivity work.
- Menu bars are optional for simple apps that are able to expose all functionality using visible buttons and toolbars. If any functionality is not visible by default, err on the side of providing a menu bar.
- Don’t display a menu bar in secondary or internal windows, like the settings dialog or file dialog. Very small main windows are likewise usually poor candidates for menu bars.
- Don’t include a menu bar in a convergent application’s mobile user interface

https://develop.kde.org/hig/components/navigation/menubar/

Based on the KDE guidelines, Jamulus should have a menu bar.

**Reply by jp8** [reply]
KDE guidelines also confirm my gut feeling about menu names:

- At the minimum, all windows should have File, Edit, Settings, and Help menus. If they apply, the window can also have View, Insert, Format, Tools and, Window menus.

**Reply by gilgongo** [reply]
> A menu bar is mandatory for applications that have a very complex command structure

Whether Jamulus should have one therefore depends on whether we think it has a "very complex command structure". I would tend to say it doesn't really though.

Also, since both menu bars and single-entry points are commonly used, I think we can assume that users will be fine with either in terms of recognition. What matters after that is usability. Is it more or less usable to have, say, an "Edit" menu that shows, say "Auto-adjust all faders" than for that to appear as part of a lager single or nested menu?

Would be good to explore both I think.

**Reply by jp8** [reply]
yes - was going by this line:

>  If any functionality is not visible by default, err on the side of providing a menu bar.

and by my gut feeling / instincts...

**Reply by gilgongo** [reply]
What is meant by "not visible by default"? Very few apps would have all functions visible without interaction, no?

**Reply by jp8** [reply]
I'm taking that to mean things like the Chat window and the Connect dialog.

But on Apple IOS and on Android, the exact same application will always have a "Primary Menu" like you suggest.  It's just a convention that on desktop OS, you would see a menubar instead.  And the convention is slowly changing, as we can see in applications like Chrome and Edge, and as the desktop and mobile OS start to converge.

Apple's guidelines are still really clear - the menubar is required.  On Windows, the menubar is still by far the standard.  On Linux really anything goes.  I think on the desktop platforms Jamulus should have the menubar, so that Mac, Windows, and Linux UI are consistent.  And on mobile (which I haven't installed yet but I guess I now need to) I suppose the Primary Menu (the hamburger thing) has submenus that exactly mirror the menubar.

**Reply by gilgongo** [reply]
The problem with a menu is that unless each top-level item is well-understood **and** the user can match that undertanding with a command they want to invoke (so, go to File to open a file, go to About to get info on the app, etc.) then they are forced to choose where to start. 

A single "door" removes this initial problem, and the user can then be left to choose from what might be almost a full list of functions in the case of Jamulus. 

![image](https://user-images.githubusercontent.com/4561747/126492354-7e063833-fb78-4be3-ba27-1c06e8c32ece.png)

**Reply by jp8** [reply]
Yes that's nice, and similar to what Chrome uses.  But note that on the Mac (at least on mine, MacOs 10.14), Chrome has this Primary menu but also a full normal menubar.



![Screenshot_2021-07-21_14-04-33](https://user-images.githubusercontent.com/6978815/126493486-2badbcf4-f4e0-4755-90ce-908701a11bfc.png)


And Firefox (below) has both primary menu and full menubar, on all platforms I think:

![Screenshot_2021-07-21_14-05-32](https://user-images.githubusercontent.com/6978815/126493742-ca943477-8079-4ddd-b192-56e2a53fb71a.png)


Did you think about how extensible it is ?  If somebody wanted to add recents or favourites, where would they appear ?

I can't see how I would change between compact and normal skin.

**Reply by gilgongo** [reply]
Note also the technique of "self labelling" here too: each section of the menu is simply grouped into things which the desigers hope are related. They don't have to put names to these groupings though.

Ineresting that Firefox uses both. Not sure how that would work if commands were in one place but not another.

**Reply by jp8** [reply]
yep I defnitely noticed that.

Chrome on MacOS uses both too.  I think you've designed the menu of the future... but are the users ready for it now ?  I'd like to hear from someone with a more modern Mac, whether the menubar is still there in Chrome.

**Reply by jp8** [reply]
Sorry I've edited my post above, after you already replied to it.

> Did you think about how extensible it is ? If somebody wanted to add recents or favourites, where would they appear ?

> I can't see how I would change between compact and normal skin.

**Reply by jp8** [reply]
And imagine you did "set faders to default level" and then wanted to undo it.  Where would the undo be ?  (Usually on the Edit menu)

**Reply by gilgongo** [reply]
In so far as we can observe that a number of common apps exhibit this "single door" technique (which has at at least a couple of advantages over standard menus), then I think we don't need to worry too much about recognition. 

Extensiblity is an interesting one, and a possible disadvantage of the approach. It would require perhaps another menu, tab within in the menu, or other "overflow" approach such as sub-grouping. BTW my example shows a sub-group in the form of "Sorting..."

For undo, I would say we could rely on Ctrl+Z since the use of undo would not be (unlike on a drawing app, say) core to use. But it might conditonally appear at the top of the menu as "Undo [last commandName]" perhaps.

**Reply by jp8** [reply]
Yeah I almost agree with you.  Chrome on MacOS has undo and redo on the Edit menu, but you can't find them anywhere on the primary menu.  It's probably why Apple is still insisting that the menubar has to be there.  Otherwise it's a step backwards to how it used to be, before "menu-driven" software - in order to use a command you had to already know it.

![Screenshot_2021-07-21_14-43-57](https://user-images.githubusercontent.com/6978815/126498920-4212da8c-6204-4e84-bef6-e6a770b9f3bb.png)

I honestly don't think it is the right step for Jamulus to take right now.  I believe on MacOS there is no option but to have the menubar.  And I believe Jamulus should have a consistent UI across all platforms, otherwise it becomes difficult to write guides and documentation.

**Reply by jp8** [reply]
I suggest we re-arrange the existing menus to File | Edit | View | Settings | Help 

Then make a future PR to add the Primary menu just like yours, and make the normal menubar default hidden on Windows and Linux.

On mobile I'm a bit unsure...  I suspect right now it has a primary menu that mirrors the current menubar.  But I don't have a build environment set up.  Are there pre-built apks ?

**Reply by ann0see** [reply]
> Are there pre-built apks ?

yes: https://github.com/jamulussoftware/jamulus/releases/tag/r3_8_0

see assets. 

And https://jamulus.io/wiki/Installation-for-Android

**Reply by jp8** [reply]
So it's as I expected - a Primary hamburger button that offers the desktop menubar.

**Comment by jp8** [comment]
Regarding the View -> Sort options, the standard guidelines suggest a couple of changes.

1 - We should change "Sort Users" to "Sort Channels".  The word "users" is tech jargon.  I think anyway I've read at one time a Jamulus style guide that tells me to call them Channels, not Users.

2 - This "no user sorting" can never make sense to users.  The channels are always sorted by something.  In this case it's according to a server-assigned slot which is an implementation detail that really should not be exposed in the UI. The server-assigned slot is meaningless to users, there is no reason for anyone to ever select it.

**Reply by gilgongo** [reply]
@jp8 Check the style guide, but I think you'd be correct about "Sort Channels":

https://jamulus.io/contribute/Style-and-Tone#standard-terms

**Reply by jp8** [reply]
thank you.  I skimmed CONTRIBUTING.md

but i did not immediately see a link to the style guide.

**Comment by jp8** [comment]
How do I make the "analyzer console" appear ?  I would like to make sure it looks ok on the menu...

**Reply by ann0see** [reply]
Look at main.cpp. I think it’s just a cli flag called analyzerconsole

**Reply by pljones** [reply]
It's an officially undocumented feature -- and shouldn't be accessible other than from the CLI, really. :)

**Comment by ghost** [comment]
jp8 wrote on july21:
> I suggest we re-arrange the existing menus to File | Edit | View | Settings | Help

I'm almost in full agreement here. What I'll suggest is:  File | Settings | Help
The Macintosh and Windows (not Unix) idea here is to use Settings and Help for
what is intuitive, and File for what remains. (Fully Unix and Linux compliant)
File (being the first menu) is Always seen, and can contain a few single depth
secondary items.
Note that File is just a meaningless label that people expect as the first item.

(KISS and minimalist)

**Reply by jp8** [reply]
Ah but I never thought of File as a meaningless label.

The first menu is *always* about the top level object you are working with.  In a word processor, the first  menu is about the document.  In a text editor, the first menu is about the text file.  In a spreadsheet program, that menu is about the spreadsheet.  In a web browser the first menu is about the web page.

The first menu is *normally* labeled File, but sometimes it is labeled after that top-level object.

The top level object in Jamulus is a mixer.  We could name the first menu Mixer if we thought it was a good idea.  When you connect to a different server, what you are really doing is connecting to a different mixer in the cloud.

I prefer to name the first menu File because it's really very standard and I don't see a strong reason to rename it.  Every little unexpected thing you do creates a bit of mental work for the user, and it all adds up.  Saving the user from unnecessary mental work is, in my view, what the keep-it-simple principle is about.

Have a look at Chrome's file menu - there is plenty that's not about files but it's all about web pages.  You can open them, close them, save them locally, print them, share them with your friends....

![Screenshot_2021-07-22_16-14-34](https://user-images.githubusercontent.com/6978815/126663797-570fa360-7ad2-4b33-9709-1d9fd71ed7b3.png)

**Comment by jp8** [comment]
> I'm almost in full agreement here. What I'll suggest is: File | Settings | Help

David - I think your approach to have everything in the File menu really closely matches the suggestion of @gilgongo for a primary menu.  You can see his screenshot in this comment:

https://github.com/jamulussoftware/jamulus/discussions/1915#discussioncomment-1032430

I would like to put all the basic functionality for musicians in a single Primary menu (often called a 'hamburger' menu because of the three horizontal layers in the icon), but with the full menubar still available for 'advanced' functions - grouping, multiple rows, loading and saving mix definitions, and so on.

I would like to make a series of pull requests for this.

Step 1 - Re-arrange existing menu items to follow industry UI guidelines (file, edit, view, settings, help).  

Step 2 - Expand new menus to add items that currently require bringing up dialogs

Step 3 - Add Primary Menu (hamburger menu) with commonly accessed functionality.

There will still be potential to discuss and agree (or disagree) the fine details in each of the pull requests.

Does the overall plan sound ok ?  Does anyone have any strong objections ?

**Reply by ghost** [reply]
The overall plan does sound ok. I won't comment further, unless the plan diverges or becomes chaotic.

**Reply by jp8** [reply]
> unless the plan diverges or becomes chaotic.

Hehe thanks for that vote of confidence.

I intend to take it slowly and get it right.  But the first pull request will be simple and quick, and that will give everyone plenty of time to try it and validate it.

**Comment by jp8** [comment]
I have opened pull request #1926

**Comment by jp8** [comment]
Now that #1926 is merged, the menu items have not changed, but a few are found on different menus.

Is there a release scheduled soon ?  There is a certain benefit in pushing UI changes gradually to the end users.  But if there is a long time before the next release, then I would like to just go ahead with the next changes.

**Reply by jp8** [reply]
I have created the second pull request #1934
